### ISSUE TITLE
with_any_role(:role1, :role2) on User Object returns Array instead of ActiveRecord::Relation

### ISSUE THREAD

**Issue by FreeApophis** [RolifyCommunity/rolify#292]
with_any_role(:role1, :role2) on User Object returns Array instead of ActiveRecord::Relation

This is a rather tiny issue, but it is rather surprising, since with_role already correctly returns a Relation Object which you can easily chain.
The problem showed up when I chained other Database scopes, and the temporary solution is to do the with_any_role at the end of the chain. This conflicts with other scopes like pagination, and of course the Array cannot be evaluated lazily.

**Comment by fabn** [comment]
The same is true for `with_all_roles`. Is there any workaround?

**Comment by EppO** [comment]
You can combine ActiveRecord relations using `.merge` that combines `WHERE` clauses using `AND` which is not helping in that case (we need to `OR` the different `WHERE` clauses). With the upcoming Rails 5 release,  we will be able to combine two ActiveRecord relation with `OR` (read this [article](http://juanitofatas.com/2015/01/31/rails-5-active-record-or/) for more info). I don't see any workaround for Rails <=4.x except playing with Arel directly.

**Comment by dkniffin** [comment]
Has there been any progress on this? I was really hopeful that this method would do exactly what I wanted, but then I saw it was an array, which prevented me from chaining it with other queries.

**Comment by dkniffin** [comment]
Alright, I figured out a way to fix this method. I've got a PR open now (#441)

**Comment by nhattan** [comment]
workaround
```ruby
# user.rb
scope :with_any_roles, ->(names) do
  joins(:roles).where("roles.name IN (?)", names)
end

# using
User = User.with_any_roles([:admin, :moderator])
```

**Comment by kakoni** [comment]
@nhattan If user has multiple roles, then with that query you will get duplicate rows.  Of course ``` User.with_any_roles([:admin, :moderator]).distinct```helps

**Comment by Mehonoshin** [comment]
Still not not released?

**Comment by nhattan** [comment]
Thanks @kakoni

**Comment by EppO** [comment]
WIP here #441 

TL;DR: Works with Rails 4.2 but not (yet) on Rails 5

**Comment by milani** [comment]
@kakoni just be aware that the workaround does not work when there is a column with type `json`. `DISTINCT` raises an exception in that case.

**Comment by milani** [comment]
@kakoni @nhattan Maybe it is safer to write it as:

```
# user.rb
scope :with_any_roles, ->(names) do
  where(id: joins(:roles).select('users.id').distinct.where("roles.name IN (?)", names))
end
```

**Comment by Kani999** [comment]
@milani - I'm using your code right now and it works like a charm.

Still waiting for this feature in a Gem

**Comment by martinstreicher** [comment]
It would be best if all role queries returned an AR, not an Array.

**Comment by NuclearMachine** [comment]
6 years 🥇
